lib/Yield/Support/SugiyamaStyleGraphLayout/NodeRanking.cpp

```diff
@@ -22,7 +22,7 @@ RankContainer rankNodes<BFS>(InternalGraph &Graph) {
   RankContainer Ranks;
 
   for (auto *Node : llvm::breadth_first(Graph.getEntryNode())) {
-    auto CurrentRank = Ranks.try_emplace(Node, Rank(-1)).first->second;
+    auto CurrentRank = Ranks.try_emplace(Node, Rank(4)).first->second;
```

Comment:
* I guess this is a workaround. If so, please add a `// TODO` about it.
* Can you describe an example that triggers this issue?
* Make a constant out of `4`

Reply:
1. It is a workaround, but consider that ranks are, in the essence, relative values. And they are only ever used internally within the layouter.

2. The issue arrives when backedges cause backedges that span before the root node (as in, go into the negative rank). This workaround doesn't fix the issue, but offsets it by 3 or 4 layers (depending on what was the initial value of the constant). Consider that in real examples I never saw more than one layer, but theoretically more is possible, so I decided to push the constant further back.

3. Sure.

Reply:
It seems like we can arbitrarily increase this value with no side effects.
If this is the case, can't we have a post-processing phase that offset all the ranks so that the lowest is 0?

Reply:
1. There is a bit of a side effect - at one point, the `layer` structure ([this one](https://github.com/revng/revng/blob/develop/lib/Yield/Support/SugiyamaStyleGraphLayout/InternalGraph.h#L194)) is created based on the ranks. And since we have 3 (or 4) extra ranks - there are 3 (or 4) extra elements of the "outside" vector created there. And that's it - no other side-effects.  
P. S. If not for this I'd have set the value to like `1 << 16` or something and would call it a day.

2. This is, in fact, what happens when layers are finalized. To remove any "holes" (as in, layers we removed when processing the graph further), the ranks of all the nodes are shifted to ensure better [logical position](https://github.com/revng/revng/blob/develop/lib/Yield/Support/SugiyamaStyleGraphLayout/InternalGraph.h#L201) relations (e.g., if we see two nodes with neighboring ranks, we can be sure they are going to end up very close vertically in the final graph).

Reply:
If we have an indexing mechanism whose "zero" is not predictable at the start, there might an opportunity to reorganize the code in terms of relative position among ranks, instead of working with absolute numbers.
Basically, drop the indexing system in favor of pointers to object representing ranks, which are part of a linked list or something similar where you can add elements in front or on the back (and maybe in the middle?).

But I don't have enough context, so please just drop a TODO of what you think it might a good long term solution to avoid the hack and let's move forward.

Reply:
I think I'm confusing you - the `Rank`s and the `Index`es have nothing to do with each other - they are separate systems.

Reply:
Yeah, sorry, I meant rank system. Ranks are "indexed", in the sense that they are associated to an absolute value.
Does this make more sense?

Reply:
`Index`es are unique node identifiers. They never change and assigned when a note is created.

`Rank`s are our way of sorting nodes, e.g. to store whether a specific node needs to be near the front of the graph, or near the back.

Ranks of all the nodes in the same horizontal layer are the same by definition. That's what we're looking for when assigning them. We're looking for sets of nodes that can be laid out in parallel to each other (on the same horizontal layer). But the relations between different ranks are purely relative: the higher the rank - the closer the node is going to be placed to the start of the graph - and that's it.

The problem with a "good long term solution" is that it would require quite a bit of a rework, and I'm not sure it's worth it - if we were to pour more time into the layouter, there are more pressing problems with it.

Reply:
Indexing of the ranks is, in its essence, an optimization - it lets us use vectors for representing the layers. And since this data-structure does impact the bottle-neck, we don't have much wiggle room about how we want to structure this.

Reply:
I'm just suggesting to leave a TODO on how to improve the situation, not to actually do it.
But yeah, let's move on.

Reply:
Yeah, I mean that if someone at some point sees this TODO and decides to "fix" this, their effort would be more useful in other parts of the layouter.

But sure, I'll leave one after all.

Reply:
Done: https://github.com/revng/revng/pull/197/commits/1b87566a41eb9a426fb56b3652a6b1ba058aee9b.